server/src/main/java/org/apache/druid/server/coordinator/CuratorLoadQueuePeon.java

```diff
@@ -183,28 +172,17 @@ public void dropSegment(
       final LoadPeonCallback callback
```

Comment:
Methods and fields in this class generally lack concurrent access documentation: https://github.com/code-review-checklists/java-concurrency#justify-document. For example, it's not clear if you removed synchronization in `dropSegment()` method because it's called only from a single thread, or because you think the new code is concurrently safe. If the latter, then `dropSegment()` seems to be a subject for https://github.com/code-review-checklists/java-concurrency#chm-race

Reply:
This class has been around forever and I agree that more could be done to document the thread safety of the class. Having said that, I don't think this PR should be focussing on it. It could possibly be addressed in https://github.com/apache/incubator-druid/projects/4. 

Looking at the callers of loadSegment() and dropSegment() I can't say for sure if they could be called by multiple threads. It doesn't seem like it from what I see. However, to better guard against future changes, I am going to slightly change the code to use putIfAbsent instead of doing the racy check-then-act.

Reply:
Why did you remove the synchronization in the first place?

The change doesn't quite work without concurrent access documentation. Concurrent access documentation should go hand-in-hand with the concurrently safe code. Without the documentation, how would anybody know why `putIfAbsent()` is used, and if that is enough?

I would not ask you to add documentation if you didn't remove synchronization and change the concurrency model of this class in the first place. But since you did, you have to be able to answer "hard" questions about this class. There is no other way to verify this PR.

Reply:
The old code was synchronizing on ```currentlyProcessing``` since it was concurrently accessed by multiple threads. Of special interest were these two places - 

https://github.com/apache/incubator-druid/blob/master/server/src/main/java/org/apache/druid/server/coordinator/CuratorLoadQueuePeon.java#L318

and 

https://github.com/apache/incubator-druid/blob/master/server/src/main/java/org/apache/druid/server/coordinator/CuratorLoadQueuePeon.java#L224

This was the key reason behind the peon assigning segments serially. It would wait for the previous load/drop request to complete before starting a new one.

Reply:
I think there should be a good balance between letting code document itself vs writing multiple lines in plain English which sometimes depending on writing style of the author doesn't always make sense to the reader.